### DISCUSSION TITLE
How to integrate spectro external codes

### DISCUSSION THREAD

**Discussion by gcanasherrera** [cloe-org/cloelib#44] [Ideas]
Discussion thread about how to integrate external spectro codes such as `effort.jl`, `comet`, `PBJ`, `pybird`. We need to assess if:

- Background + Perturbations split in `cloelite` is good enough to accomodate the codes
- How to deal with the computation of the Pk

List of minimal ingredients needed to make the EFTofLSS computation (think on physical terms, not computationally):

- Addition of nuisance parameters (which model parametrisation do we want to use? For the moment I would say to use the same one we already had, i.e. the PBJ one), with interface to make them selectable by the user
- Integration of external codes spectro codes (need to make them ready to receive an external background). This in principle does much of the job that was previously done in the eft module
-  Revise the way the growth rate is defined in cosmologies with massive neutrinos. So far we are using the large-scale limit (i.e. without k-dependence), but we measure it as fs8/s8, which should not be the way to go. That's because s8 is affected by the presence of neutrinos and the ratio fs8/s8 does not represent the large-scale limit. One better thing to do to get this large-scale approximation would be to compute $`\lim_{k\rightarrow 0}\frac{P_{dv}(k)}{P_{dd}(k)}`$
- ...

**Comment by gcanasherrera** [comment]
@cloe-library/cloe-maintainers let's comment here

**Comment by marcobonici** [comment]
Hi @gcanasherrera ,
We might need to think a bit more about it, but I think this should be good enough.

I also would love to listen from the other people involved, but my opinion is that:

1. External codes we are gonna wrap (like the ones you names) should be in charge of doing the most complex calculations that can have a different implementations (e.g. computing the multipoles of the power spectrum) 
2.  `cloe` should be in charge of the observational effects (AP and mask)

Although all the previous codes can apply the observational effects, I think it would be beneficial to have a single implementation (especially for the AP!) thoroughly tested and highly efficient (I am not worried about the window convolution, because this can be written as a matrix-vector multiplication).

If we agree on this, we should discuss a common API. My idea is that we need to pass a linear perturbation theory object, since the only thing we need to pass to `pbj` or `pybird` is the linear Pmm (for emulators, we don't even need that).

Lemme know which are your thoughts! :)

**Reply by gcanasherrera** [reply]
Thanks @marcobonici, I haven't played with those codes, so I think @AndreaPezzotta @chiaramoretti are our experts for this.

**Reply by santiagocasas** [reply]
thanks @marcobonici ! if we do the AP on this side (which makes sense) then it would be good for all the codes to use the same background quantities, right? So we should also pass a background object, no?

**Reply by gcanasherrera** [reply]
That would be ideal, but we need to make sure that all those codes still keep their functionality to be CLOE-independent, somehow, otherwise it could be counterproductive. If so, we will be basically asking the developers of those codes to potentially add some funcitonalities to make it CLOE background object compatible, right?

**Reply by ntessore** [reply]
Someone should come up with an API for cosmology objects!

**Reply by gcanasherrera** [reply]
We accept PRs from collaborators working in external projects too. I wonder if the `cosmology.api` people would be happy to give us a hand on this...?

**Reply by AndreaPezzotta** [reply]
Apologies for being late! I pretty much agree with the points that were raised in this thread. I think having a single background class that is internal to cloe and that can be propagated to all the other modules comes with some evident benefits. Most trivially, we would have much more control of potential bugs if we had a single background to deal with, and if the rest of the code is prepared to receive the background as input it would also be simpler to implement extensions to non-standard models.

One point that we have to think about is how much effort we need to ask to the developers of the external code to make them CLOE-compatible. I understand that if it is PBJ, effort, or comet we are basically talking about us, but maybe this won't always be the case. As for comet, I can say that it should be definitely possible to use background quantities other than the ones that are currently wrapped in the library. It won't come for free though, so I will have to add modifications to allow for this extra degree of freedom.

Happy to discuss further at our next call!

**Reply by marcobonici** [reply]
Sure!
Here in particular I was referring to the implementation of the AP and the computation of Hubble and angular diameter distance.
My idea is to have something like

```
Pl_0, Pl_2, Pl_4 = GC_backend.get_Pl_multipoles(...)
```

Where `GC_backend` is one of the codes that can compute the clustering multipoles.
After this, `cloe` should be able to apply the AP without using the external codes.

```
Pl_0_AP, Pl_2_AP, Pl_4_AP = cloe.apply_AP(Pl_0, Pl_2, Pl_4, background_quantities).
```

I don't  think that should be too difficult as all the codes can compute the multipoles without the AP.

**Reply by martincrocce** [reply]
Don't you need more than \ell=4 as input to generate \ell=4 w/ AP as output ? @AndreaPezzotta can comment maybe.

**Reply by AndreaPezzotta** [reply]
Yes, there is a non-negligible component that comes from higher-order multipoles. I believe this could be fixed by retrieving the 2d P(k,mu) from the individual codes, since in that case the information from higher-order terms is present. 

For comet, this is anyway approximated, since we are getting contributions only up to ell=6, which is anyway a reasonable cut. But even that, we would then need to reconstruct the 2d power spectrum also including the ell=6 terms, and pass it to cloe. This is happening already in comet, but I need to check if we need extra coding, such an appropriate interface, when using it in cloe.

**Reply by AndreaPezzotta** [reply]
Btw it would also be interesting to see if the residual contributions from higher-order multipoles make a difference in terms of cosmological parameters, for example for DR1

**Reply by chiaramoretti** [reply]
I was about to comment on this as well...for PBJ we do the AP internally on P(kmu) and then output the multipoles already including AP, which of course can be turned off but then you need the higher order multipole contributions. One option is outputting the P(kmu) instead of the multipoles, I can definitely add a function.
Notice that the same is true for the post-recon BAO power spectrum that we have implemented.

In general I agree on having a shared background module/computations, though we need to be a bit careful for non-standard situations: I'm thinking for example we want to fit some exotic model for which we have the correct background computations but then use them jointly with a nonlinear code that does not treat nonlinearities properly for that model (let's say e.g. f(R), and the nonlinear code does not include the proper scale dependent growth functions)

**Reply by AndreaPezzotta** [reply]
One more thing to think about is how to properly implement analytical marginalisation for these models. To do that, we need to retrieve not just the final power spectrum multipoles but rather the individual building blocks, already separated into the different combinations of bias, noise, and counterterm parameters. I think that returning the whole P(k,mu) is not enough in this case

**Reply by gcanasherrera** [reply]
Naive question but, couldn't the analytical marginalisation happen internally directly in PBJ, for instance?

**Reply by ntessore** [reply]
> In general I agree on having a shared background module/computations, though we need to be a bit careful for non-standard situations

This is a very good comment, and if one is not massively careful, it can also bite you the other way: A non-standard cosmology affects things nominally computed using only the background cosmology, e.g., the naive lensing weights.

We thought a lot about how to best deal with this when we were fiddling with our cosmology API implementation, but we ultimately didn't find a good solution.

**Reply by chiaramoretti** [reply]
Right, so maybe another option is to make the P_ell functions dependent on the alphas as parameters that are then computed in the shared background module? The way I have it now I can pass the alphas as nuisance parameters, and then use them to compute the distorted k,mu grid and amplitude rescaling to use to construct the model, and the function outputs a [P_ell] list, and a [P_marg] list, the first containing the multipoles, the second the multipoles for each marginalised term

This has some limitations (can only marginalise over all possible parameters, or none), but might be a good option? alphas computed outside and passed as parameters

In any case I think we'll need to have two versions of the likelihood, one with and one without analytic marg?

**Reply by chiaramoretti** [reply]
> > In general I agree on having a shared background module/computations, though we need to be a bit careful for non-standard situations
> 
> This is a very good comment, and if one is not massively careful, it can also bite you the other way: A non-standard cosmology affects things nominally computed using only the background cosmology, e.g., the naive lensing weights.
> 
> We thought a lot about how to best deal with this when we were fiddling with our cosmology API implementation, but we ultimately didn't find a good solution.

This maybe goes in the "nonstandard cosmologies" section of the discussion, but one option might be to put warnings/breaks, e.g. my spectro code works for LCDM, neutrinos, wCDM, w0waCDM, so if I ask for a different model the code stops. We probably need a higher level discussion where we decide how to specify the cosmo model we want to fit -- either by passing the parameters, or adding a flag 'LCDM', 'wCDM' etc? I see limitations for both options at the moment

**Reply by chiaramoretti** [reply]
> Naive question but, couldn't the analytical marginalisation happen internally directly in PBJ, for instance?

the modelling yes, but you need to change the likelihood to add the analytically marginalised contribution to the 'standard' computation

**Reply by ntessore** [reply]
> This maybe goes in the "nonstandard cosmologies" section of the discussion, but one option might be to put warnings/breaks, e.g. my spectro code works for LCDM, neutrinos, wCDM, w0waCDM, so if I ask for a different model the code stops. We probably need a higher level discussion where we decide how to specify the cosmo model we want to fit -- either by passing the parameters, or adding a flag 'LCDM', 'wCDM' etc? I see limitations for both options at the moment

I am seeing the problem more on the code side itself: I am imagining someone codes up the lensing weight

```py
def q(zlens: float, zsrc: float, cosmo: Background) -> float:
    x_l = cosmo.angular_diameter_distance(zlens)
    x_s = cosmo.angular_diameter_distance(zsrc)
    x_ls =  cosmo.angular_diameter_distance(zlens, zsrc)
    return x_l * x_ls / x_s
```

How could this code know that it needs to factor in a correction for a non-trivial Poisson equation when `cosmo` describes a non-standard cosmology?

**Reply by ntessore** [reply]
To explain why I am thinking of this, in GLASS we do have exactly such a function that computes the convergence from matter shells. To answer my own question, I guess that code would have to (optionally) accept the perturbations object and, if given, run a "does this need non-standard lensing?" check on it.

```py
def q(zlens: float, zsrc: float, cosmo: Background | Perturbations) -> float:
    x_l = cosmo.angular_diameter_distance(zlens)
    x_s = cosmo.angular_diameter_distance(zsrc)
    x_ls =  cosmo.angular_diameter_distance(zlens, zsrc)
    if not hasattr(cosmo, "non_standard_lensing_flag_or_whatever", False):
        # standard lensing sector
        return x_l * x_ls / x_s
    # non-standard lensing sector
    ...
```

**Reply by marcobonici** [reply]
@ntessore , thanks for discussing this, which I think is important, but this is OT.
Should we open a new discussion related to this topic?

My general idea is that we should not aim to have a code that is able to make coffee (abstraction and generality comes at a cost, coding wise). So, I think we should ask ourselves "Which are the science case we might want to study?". This is a scientific, not a coding question. Once we have this answer, we should ask "Which is the most general way we can account for differences among the different models?" Besides the calculations of the linear and non linear Pk (that is not problematic, because the code will just deliver a different array), I am mostly worried by non-standard lensing factor as Nicolas said. In that case, my initial guess is that the `cosmology` class should contain a `lensing_factor` method. In general cosmologies (LCDM, w0waCDM, etc...) it is just what Nicolas wrote. In more general theories with a different function, we could modify that function.
This is not necessarily the best possible idea, so please point out problems with it.

**Reply by gcanasherrera** [reply]
I would invite to transfer the discussion about a necessary common cosmology API to a new issue, and try our best to make it happen. The future will thank us.

In the meanwhile, we can keep the discussion here for the specific spectro codes' needs. Do we agree team? 👍

**Comment by marcobonici** [comment]
Regarding AP, up to now I saw two different approaches:
1. The PT code gives you the power spectrum multipole (so no \mu dependency). In order to apply the AP, you sum the multipoles after multiplying by the Legendre multipoles and then you project back. This is the strategy followed by `pybird`.
2. The PT code gives you the P(k, \mu). In order to apply the AP, we just project (after applying the corrections). This is the approach followed by `velocileptors`, `PBJ` and `comet` (@chiaramoretti @AndreaPezzotta feel free to correct me).

This shouldn't matter a lot; there was a DESI [paper](https://arxiv.org/abs/2404.07272) showing a comparison of `pybird` and `velocileptors`. The agreement is excellent, as showin in this plot (this is for LRG, with a volume of 8 Gpc^3).

![image](https://github.com/user-attachments/assets/dbd6c7b2-3f92-4c3a-a02c-4b9ab67353c0)

Here for instance the AP was not introducing any relevant difference (and of course the codes have other differences).

If we think it is really important to follow approach 2, it is fine. With `PBJ` we just retrieve the wedges and dol our stuff. With `pybird`, we can have an intermediate step where `pybird` computes the multipoles and our wrapper compose them to reconstruct the wedges.

**Reply by martincrocce** [reply]
Whatever enables analytical marginalisation as currently done in PBJ and COMET to avoid extra work in that side no ? (and we need to change the likelihood).

**Reply by Alkistis** [reply]
I am thinking Option 2 is a bit cleaner and gives an easier way to the user to implement a wedges analysis with excluded regions of \mu as in [this BOSS paper](https://arxiv.org/abs/1607.03147).

**Reply by marcobonici** [reply]
(Being able to discuss with you all is something that drives me to tears, thanks.)

Then let's go with option 2!
The API then might look like

```
Pkmu = GC_backend.get_wedges(...)
Pl_0, Pl_2, Pl_4 = cloe.apply_AP(Pkmu, GC_backend)
data_vector = cloe.window_convolve(Pl_0, Pl_2, Pl_4, window)
```

This looks good to me. The other step we need to discuss is analytical marginalization.
While for the former steps (AP and window) we might want to implement a single version of it in `cloe` and use it for all the GC backends, I think for the analytical marginalization we might want to use a different approach.
This has already been implemented in `PBJ` and `comet`, right? Which are your thoughts, @chiaramoretti @AndreaPezzotta ?

**Reply by martincrocce** [reply]
We need to check w/ Raul if BACCO can do 2. Are @chiaramoretti and @AndreaPezzotta ok with opt 2 ?

**Reply by marcobonici** [reply]
It should, for sure. Even if BACCO computes the multipoles, rather than the wedges, we can always sum them to get the wedges (this is what `pybird` does).
In that case, we would add a step in the wrapper, doing this ourselves.

**Reply by marcobonici** [reply]
To be concrete, in that case we would define

```
def cloe.sum_Legendre_multipoles(Pl_0, Pl_2, Pl_4):
    #sumeverything together and return wedges
```

Then, in `BACCO`, we would do

```
def BACCO.get_wedges(...):
    Pl_0, Pl_2, Pl_4 = BACO.get_Pl(...)
    return sum_Legendre_multipoles(pl_0, Pl_2, Pl_4)
```

**Reply by chiaramoretti** [reply]
I think if we go with option 2 for PBJ it could go like this:
* no analytic marg: I add a function that gives the P(kmu), cloe applies AP and projects to multipoles
* analytic marg: as above but I return P(kmu) for the non-marg part, plut the P(kmu) terms that need to be marginalised over. cloe applies AP to each and projects each to multipoles

The likelihood function must have two options, for no analytic marg it does the usual (Ptheory-Pdata) inv cov (Ptheory-Pdata), for the analytic marg version it does the same (but theory is different because some terms are not included) then sums the analytically marginalised part

Apologies for the pseudo-math, I hope it's still readable :)

**Reply by gcanasherrera** [reply]
To my gc expert fellows, could you please update the description of the discussion above with the minimum ingredients we need to compute EFTofLSS? Thanks!

**Reply by chiaramoretti** [reply]
Listing them here for PBJ specifically, we can update the common ones when we have details from other codes:
* cosmology: h, Obh2, Och2, ns, As, Tcmb, z (used to compute the Plinear and for IR resummation; also supports Mnu, w0,wa)
* if we want to compute Plinear externally, we need to pass it to the functions computing the loop corrections
Both ingredients above can be passed in a common `cosmology` dictionary

* additional options: kind of IRresummation (Eisenstein-Hu or Discrete Sine Transform), if we want to perform redshift rescaling with the growth functions, if we want to do AP and window convolution (can be turned off if we do it externally)
* the "final" function that computes the multipoles/Pkmu also wants: nuisance parameters (b1, b2, bG2, bG3, c0, c2, c4, aP, e0k2, e2k2 -- same as CLOE), systematics parameters (for now, f_out and sigma_z -- TBD for interlopers), nbar, the grid of k on which to compute the multipoles/Pkmu, growth functions f, D (optional, they can also be computed internally from the cosmology dict)
I can go into more technical details once we move to [#48 ](https://github.com/cloe-library/cloelite/issues/48)

**Reply by chiaramoretti** [reply]
To keep the background/linear computation external:
* compute Plinear externally and pass it to the relevant pbj function in a dictionary with the cosmology;
* compute growth functions externally and pass them to the Pkmu/multipoles function
* for AP:
    * either I return Pkmu and we apply it externally
    * or we compute the alphas and pass them to the function computing the multipoles that takes care of distorting the k,mu grids and rescaling the power spectrum amplitude, this option is already supported
